sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala

```diff
@@ -1312,10 +1326,24 @@ class Dataset[T] private[sql](
         colRegex(colName)
       } else {
         val expr = resolve(colName)
-        Column(expr)
+        Column(addDataFrameIdToCol(expr))
```

Comment:
This is the only place that requires calling `addDataFrameIdToCol`. How about colRegex? 

We might need a clear comment and explain which cases we should call it?

Reply:
Same question: doesn't it apply to `colRegex`?

Reply:
good catch! We should apply it in `colRegex` as well.